core/src/main/scala/filodb.core/query/QueryContext.scala

```diff
@@ -22,26 +22,50 @@ case class PromQlQueryParams(promQl: String, startSecs: Long, stepSecs: Long, en
 
 case object UnavailablePromQlQueryParams extends TsdbQueryParams
 
+case class IndividualQuota(
+        execPlanSamples: Int = 1000000,       // Limit on ExecPlan results in samples, default is 100K
+        execPlanResultBytes: Long = 18000000, // Limit on ExecPlan results in bytes, default is 18MB
+        groupByCardinality: Int = 100000,     // Limit on "group by" clause results, default is 100K
+        joinQueryCardinality: Int = 100000,   // Limit on binary join results, default is 100K
+        timeSeriesSamplesScanBytes: Long = 300000000)  // Limit on max data scanned per shard, default is 300 MB
+
+object IndividualQuota {
+  def apply(warnDefaults: Boolean): IndividualQuota = {
+    if (warnDefaults) {
+      IndividualQuota(
+        execPlanSamples = 50000,
+        execPlanResultBytes = 15000000,
+        groupByCardinality = 50000,
+        joinQueryCardinality = 50000,
+        timeSeriesSamplesScanBytes = 150000
+      )
+    } else {
+      IndividualQuota()
+    }
+  }
+}
 case class PlannerParams(applicationId: String = "filodb",
                          spread: Option[Int] = None,
                          spreadOverride: Option[SpreadProvider] = None,
                          shardOverrides: Option[Seq[Int]] = None,
                          targetSchemaProviderOverride: Option[TargetSchemaProvider] = None,
                          queryTimeoutMillis: Int = 60000, // set default to match default http-request-timeout
-                         sampleLimit: Int = 1000000,
-                         groupByCardLimit: Int = 100000,
-                         joinQueryCardLimit: Int = 100000,
-                         resultByteLimit: Long = 18000000,  // 18MB
+                         enforcedQuota: IndividualQuota = IndividualQuota(),
+                         warnQuota: IndividualQuota = IndividualQuota.apply(warnDefaults = true),
+                         queryOrigin: Option[String] = None, // alert/dashboard/rr/api/etc
+                         queryOriginId: Option[String] = None, // an ID of rr/alert
+                         queryPrincipal: Option[String] = None, // user, entity initiating query
                          timeSplitEnabled: Boolean = false,
                          minTimeRangeForSplitMs: Long = 1.day.toMillis,
                          splitSizeMs: Long = 1.day.toMillis,
                          skipAggregatePresent: Boolean = false,
                          processFailure: Boolean = true,
                          processMultiPartition: Boolean = false,
                          allowPartialResults: Boolean = false)
+
 object PlannerParams {
   def apply(constSpread: Option[SpreadProvider], sampleLimit: Int): PlannerParams =
-    PlannerParams(spreadOverride = constSpread, sampleLimit = sampleLimit)
+    PlannerParams(spreadOverride = constSpread, enforcedQuota = IndividualQuota(execPlanSamples = sampleLimit))
```

Comment:
I'm not sure where we use this specific `apply`, but it might construct `PlannerParams` with some warning limits less than the enforced limits.

Reply:
this is expected, isn't it? the warning limits are smaller than enforced, ie at 2MB I warn you, at 10MB I won't let your query run.